File: backend/app/schemas/notification.py
from datetime import datetime

from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

from app.database.models.notification import NotificationType


class NotificationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type: NotificationType
    title: str
    message: str
    link: str | None = None
    is_read: bool
    created_at: datetime


class UnreadNotificationCountResponse(BaseModel):
    unread_count: int


class NotificationSnapshotResponse(BaseModel):
    items: list[NotificationResponse]
    unread_count: int
    unread_message_count: int


class NotificationLinkReadRequest(BaseModel):
    link: str = Field(min_length=1, max_length=255)


class NotificationLinkReadResponse(BaseModel):
    marked_count: int
    unread_count: int
    unread_message_count: int


class NotificationPreferences(BaseModel):
    in_app_enabled: bool = True
    email_enabled: bool = True
    messages: bool = True
    connections: bool = True
    communities: bool = True
    applications: bool = True
    email_frequency: Literal["immediate", "daily", "never"] = "immediate"
    quiet_hours_enabled: bool = False
    quiet_hours_start: str = Field(default="22:00", pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    quiet_hours_end: str = Field(default="08:00", pattern=r"^([01]\d|2[0-3]):[0-5]\d$")

File: frontend/src/types/notification.ts
export type NotificationType =
  | "COMMUNITY_APPLICATION_RECEIVED"
  | "COMMUNITY_APPLICATION_ACCEPTED"
  | "COMMUNITY_APPLICATION_REJECTED"
  | "COMMUNITY_INVITATION_ACCEPTED"
  | "COMMUNITY_INVITATION_RECEIVED"
  | "CONNECTION_REQUEST_RECEIVED"
  | "CONNECTION_REQUEST_ACCEPTED"
  | "COMMUNITY_MEMBER_JOINED"
  | "COMMUNITY_MEMBER_LEFT"
  | "COMMUNITY_MEMBER_REMOVED"
  | "COMMUNITY_OWNERSHIP_TRANSFERRED"
  | "PRIVATE_MESSAGE_RECEIVED";

export interface AppNotification {
  id: number;
  type: NotificationType;
  title: string;
  message: string;
  link: string | null;
  is_read: boolean;
  created_at: string;
}

export interface NotificationSnapshot {
  items: AppNotification[];
  unread_count: number;
  unread_message_count: number;
}

export interface NotificationPreferences {
  in_app_enabled: boolean;
  email_enabled: boolean;
  messages: boolean;
  connections: boolean;
  communities: boolean;
  applications: boolean;
  email_frequency: "immediate" | "daily" | "never";
  quiet_hours_enabled: boolean;
  quiet_hours_start: string;
  quiet_hours_end: string;
}
